client/client.go

```diff
@@ -12,63 +12,118 @@
+with an implementation of client.AuthData. The attribute names should be documented with their return types and
+considered part of the public API for the authenticator.
+
+Receivers are responsible for obtaining a client.Info from the current context and enhancing the client.Info
+with the net.Addr from the peer. For HTTP requests, this is typically the IP address of the client. This new context
+should then be propagated down the pipeline when calling the next consumer. Typically, however, receivers would delegate
+this processing to helpers such as the confighttp or configgrpc packages: both contain interceptors that will enhance
+the context with the client.Info, such that no actions are needed by receivers that are built using confighttp.HTTPServerSettings
+or configgrpc.GRPCServerSettings.
+
+Consumers
+
+Processors and exporters have access to the client.Info via client.FromContext. Among other usages, this data can be used to:
+
+- annotate data points with authentication data (username, tenant, ...)
+
+- route data points based on authentication data
+
+- rate limit client calls based on IP addresses
+
+Neither processors nor exporters are expected to change data from the client.Info contained in the context, although
+there might be reasons to do so. In that case, this should be clearly documented as part of the component's README file.
+
+Processors and exporters relying on the existence of data from the client.Info, especially client.AuthData, should
+clearly document this as part of the component's README file. The expected pattern for consuming data is to allow
+users to specify the attribute name to use in the component. The expected data type should also be communicated to
+users, who should then compare this with the authenticators that are part of the pipeline.
+
+For example, assuming that the OIDC authenticator pushes a subject string attribute and that we have a hypothetical
+authprinter processor that prints the username to the console, this is how an OpenTelemetry Collector configuration
+would look like:
+
+  extensions:
+    oidc:
+      issuer_url: http://localhost:8080/auth/realms/opentelemetry
+      audience: collector
+
+  receivers:
+    otlp:
+      protocols:
+        grpc:
+          auth:
+            authenticator: oidc
+
+  processors:
+    authprinter:
+      attribute: subject
+
+  exporters:
+    logging:
+
+  service:
+    extensions: [oidc]
+    pipelines:
+      traces:
+        receivers: [otlp]
+        processors: [authprinter]
+        exporters: [logging]
+*/
 package client // import "go.opentelemetry.io/collector/client"
 
 import (
 	"context"
 	"net"
-	"net/http"
-
-	"google.golang.org/grpc/peer"
 )
 
 type ctxKey struct{}
 
-// Client represents a generic client that sends data to any receiver supported by the OT receiver
-type Client struct {
-	IP string
-}
+// Info contains data related to the clients connecting to receivers.
+type Info struct {
```

Comment:
We need to make this struct immutable, otherwise we have troubles in multi-pipeline environment.

Reply:
I read your example and understood how we can generate undefined behaviors.